components/webdriver_server/lib.rs

```diff
@@ -110,6 +113,55 @@ pub fn start_server(port: u16, constellation_chan: Sender<ConstellationMsg>) {
         .expect("Thread spawning failed");
 }
 
+struct KeyInputState {
+    pressed: HashSet<String>,
+    alt: bool,
+    shift: bool,
+    ctrl: bool,
+    meta: bool,
+}
+
+impl KeyInputState {
+    pub fn new() -> KeyInputState {
+        KeyInputState {
+            pressed: HashSet::new(),
+            alt: false,
+            shift: false,
+            ctrl: false,
+            meta: false,
+        }
+    }
+}
+
+struct PointerInputState {
+    subtype: PointerType,
+    pressed: HashSet<u64>,
+    x: u64,
+    y: u64,
+}
+
+impl PointerInputState {
+    pub fn new(subtype: &PointerType) -> PointerInputState {
+        PointerInputState {
+            subtype: match subtype {
+                PointerType::Mouse => PointerType::Mouse,
+                PointerType::Pen => PointerType::Pen,
+                PointerType::Touch => PointerType::Touch,
```

Comment:
What is the purpose of this? Why not accept a subtype by value instead?

Edit: I see that PointerType doesn't implement Clone or Copy. I recommend making a patch to https://hg.mozilla.org/mozilla-central/file/tip/testing/webdriver/src/actions.rs to fix this. https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Source_Code/Mercurial and https://moz-conduit.readthedocs.io/en/latest/phabricator-user.html have some steps describing how to do that.

Reply:
Will do that!

Reply:
https://phabricator.services.mozilla.com/D39110